Add Error Handling To Rust-verkle

ControlCplusControlV opened the pull request.

ControlCplusControlV commented:
This PR fixes a lot of unsafe unwraps, I'm not totally happy with where it's at so happy to fix some things, but wanted to get *something* in before it got too stale, this requires a lot of cascading changes so I figured going little by little may be a better approach (maybe make a new milestone with different sub-issues for each core type that needs `unwrap` safety)

kevaundray commented on verkle-trie/src/database.rs:
```diff
@@ -46,6 +46,7 @@ pub trait WriteOnlyHigherDb {
     fn insert_branch(&mut self, key: Vec<u8>, meta: BranchMeta, _depth: u8) -> Option<BranchMeta>;
 }
 
+
```
Is this because of cargo fmt?

ControlCplusControlV replied:
yes

kevaundray commented on verkle-trie/src/database/generic.rs:
```diff
@@ -30,7 +31,8 @@ impl<T: BatchWriter> WriteOnlyHigherDb for GenericBatchWriter<T> {
         labelled_key.extend_from_slice(&key);
-        self.inner.batch_put(&labelled_key, &meta.to_bytes());
+        self.inner
+            .batch_put(&labelled_key, &meta.to_bytes().unwrap());
```
Can you explain the unwrap seen here? I'm guessing its an incremental change and will be eventually removed also?

ControlCplusControlV replied:
Yes, so because all of our conversions to bytes are fallible (and shouldn't logically be ignored imo, since they can sometimes be handled, the app should fail though not the library) these unwraps must slowly move upward until they can make it to the consumer

kevaundray commented on verkle-trie/src/database/generic.rs:
```diff
@@ -108,7 +111,7 @@ impl<T: BareMetalKVDb> ReadOnlyHigherDb for GenericBatchDB<T> {
         self.inner
             .fetch(&labelled_key)
-            .map(|old_val_bytes| StemMeta::from_bytes(&old_val_bytes))
+            .map(|old_val_bytes| StemMeta::from_bytes(old_val_bytes.try_into().unwrap()).unwrap())
```
Same question here for unwrap.

ControlCplusControlV replied:
Byte conversions are fallible, so we unfortunately must have unwraps for them, which I plan on slowly moving upwards through `Result`s

kevaundray commented on verkle-trie/src/database/meta.rs:
```diff
@@ -11,86 +12,89 @@ pub struct StemMeta {
-        assert_eq!(len, 64 * 3 + 32 * 3);
-
-        use ark_serialize::CanonicalDeserialize;
+        if !(len == 64 * 3 + 32 * 3) {
```
This if statement seems redundant, if bytes is an array with its type defined to be 288.

Since we are putting the length outside of the function, a minor note for another PR is to explain where the 288 comes from so it doesn't look like a magic number

ControlCplusControlV replied:
Had to convert it back to a `Vec` for current type constraints, will add a `TODO` though

kevaundray replied:
Not sure I understand where vector comes into play here, the length of that array is always 288.

ControlCplusControlV replied:
It's an ugly artifact of on-going trait refactor, I tried to make a generic `ToBytes` trait which used Generic slices but didn't work in every instance so for now Generic Slices got moved to a `Vec` until that refactoring can be more cleanly added

kevaundray commented on verkle-trie/src/database/meta.rs:
```diff
@@ -123,35 +125,41 @@ impl BranchMeta {
+impl FromBytes<[u8; 96]> for BranchMeta {
+    fn from_bytes(bytes: [u8; 96]) -> Result<BranchMeta, SerializationError> {
+        let len = bytes.len();
+        if !len == 32 + 64 {
```
Same here again, since bytes is not a slice

ControlCplusControlV replied:
Wrong branch 😨 , will fix this, some issues required this to go back to just a slice when I was working with it

kevaundray commented on verkle-trie/src/from_to_bytes.rs:
```diff
@@ -1,36 +1,46 @@
-        CanonicalDeserialize::deserialize(bytes).unwrap()
+impl FromBytes<&[u8]> for Element {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, SerializationError> {
+        let result = CanonicalDeserialize::deserialize(bytes)?;
```
Could probably do `Ok(CanonicalDeserialize::deserialize(bytes)?)` since the variable name `result` is not providing anything helpful here

kevaundray commented on verkle-trie/src/from_to_bytes.rs:
```diff
@@ -1,36 +1,46 @@
         let mut bytes = [0u8; 32];
-        self.serialize(&mut bytes[..]).unwrap();
-        bytes.to_vec()
+        let result = self.serialize(&mut bytes[..])?;
```
Is result here unused?

ControlCplusControlV replied:
Nope, removed

kevaundray commented on verkle-trie/src/proof/stateless_updater.rs:
```diff
@@ -525,10 +538,10 @@ impl SparseVerkleTree {
-            // then it will be the old commitment
-            // If we have then it will be in updated commitments
+            let child_index = prefix.pop().unwrap(); // Safety: Fine unwrap because we've checked prefix isn't empty
+                                                     // If we have never updated the parent node before,
```
Why was the comment moved here?

ControlCplusControlV replied:
Not sure, some weird format thing, moved it back

kevaundray commented:
Hi, thanks for the PR!

I'd advise modifying the relevant issue to put milestones down and break the PR up, modifying parts of the code at a time as this PR will be quite big.

Let me know if you need help with this

ControlCplusControlV commented:
I defined some future work in #62 of what I think still needs to be done. Lmk if you think anything should be added/changed, also addressed comments for this PR. I addressed comments as well, hopefully code should be better now, sorry I pushed from wrong device so branch was up to date

kevaundray commented:
Np, will re-review this tomorrow

kevaundray commented on verkle-trie/src/database/meta.rs:
```diff
@@ -11,86 +12,102 @@ pub struct StemMeta {
-                &hex::encode(scalar_to_array(&self.hash_stem_commitment)),
+                &hex::encode(scalar_to_array(&self.hash_stem_commitment).unwrap()),
             )
             .finish()
     }
 }
 
-fn point_to_array(p: &Element) -> [u8; 64] {
+use ark_serialize::CanonicalSerialize;
+
+fn point_to_array(p: &Element) -> Result<[u8; 64], SerializationError> {
     let mut bytes = [0u8; 64];
-    use ark_serialize::CanonicalSerialize;
-    p.serialize_uncompressed(&mut bytes[..]).unwrap();
-    bytes
+    p.serialize_uncompressed(&mut bytes[..])?;
```
I'm somewhat a bit uncertain about this because if the library is no longer able to serialise a point, then theres no recovery from that and the node should stop immediately. There is no clear decision on whether this should indeed propagate an error, so I'll leave it as is since it makes it usable in other non-node software whom may never want to panic

ControlCplusControlV replied:
My main thought here had been similar, that if serialization fails its unrecoverable, but I think we might also want to consider how the initial "Verge" migration might have a scenario where someone's node is migrating their entire DB in the background, hits an out of space issue, and the node might either want to more aggressively delete old Merkle content, or atleast safely halt so that progress isn't lost

kevaundray commented on verkle-trie/src/proof.rs:
```diff
@@ -261,10 +261,18 @@ impl std::fmt::Display for VerkleProof {
             let mut comm_serialised = [0u8; 32];
-            comm.serialize(&mut comm_serialised[..]);
-            hex::encode(comm_serialised)
+            let result = comm.serialize(&mut comm_serialised[..]);
```
Can do:

```rust
             match comm.serialize(&mut comm_serialised[..]) {
                 Err(_) => return Err(std::fmt::Result::Err(std::fmt::Error)),
                 Ok(_) => Ok(hex::encode(comm_serialised)),
             }
```

kevaundray commented on verkle-trie/src/proof.rs:
```diff
@@ -261,10 +261,18 @@ impl std::fmt::Display for VerkleProof {
         }) {
-            write!(f, "{} ", comm)?;
+            let output;
+            match comm {
```
Can do:

```rust
             let output = match comm {
                 Err(_) => return Err(std::fmt::Error),
                 Ok(v) => v,
             }

```

kevaundray commented on verkle-trie/src/proof.rs:
```diff
@@ -359,12 +366,12 @@ mod test {
         }
         let root = vec![];
-        let meta = trie.storage.get_branch_meta(&root).unwrap();
+        let _meta = trie.storage.get_branch_meta(&root).unwrap();
```
Why was this prepended with an underscrore?

ControlCplusControlV replied:
This is in a test attempting to make sure this call executes correctly, but the result isn't needed, will remove the variable name but we don't care about the result so much as we care that the last `.unwrap()` didn't fail

kevaundray commented on verkle-trie/src/proof/stateless_updater.rs:
```diff
@@ -35,14 +36,21 @@ pub(crate) fn update_root<C: Committer>(
-    assert_eq!(keys.len(), updated_values.len());
+) -> Result<Element, VerkleError> {
+    if !values.len() == updated_values.len() {
+        return Err(VerkleError::UnexpectedUpdatedLength);
```
Can you modify the `UnexpectedUpdatedLength` variant to take also take in two integers; values.len and updated_values.len ?

This means that error messages can become more descriptive in the future

kevaundray commented on verkle-trie/src/proof/stateless_updater.rs:
```diff
@@ -35,14 +36,21 @@ pub(crate) fn update_root<C: Committer>(
+        return Err(VerkleError::UnexpectedUpdatedLength);
+    }
+    if !keys.len() == updated_values.len() {
+        return Err(VerkleError::MismatchedKeyLength);
```
Same comment as UnexpectedUpdatedLength

kevaundray commented:
Generally looks good to me, some comments regarding:

- Prepending variables with an unserscore; we can remove them if they are not being used and clippy is complaining.

- Adding extra information into the error variants; this can be shunted to another PR though so we can merge this one ASAP

ControlCplusControlV commented:
CI is being weird, but enabled linting again

kevaundray reviewed (approved):
Thanks!
